[libpas] Extend Retag-on-Scavenge to cover bitfit objects as well

Achierius opened the pull request:
#### 742c2f8c7a2a8d9104b1a803162fc1a3f66214b3
<pre>
[libpas] Extend Retag-on-Scavenge to cover bitfit objects as well
<a href="https://rdar.apple.com/156723433">rdar://156723433</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300151">https://bugs.webkit.org/show_bug.cgi?id=300151</a>

Reviewed by Dan Hecht.

This patch finishes the retag-on-scavenge work by extending the
implementation to cover bitfit objects as well.
The main issue with the prior implementation is that it would attempt
to retag even for calls to pas_bitfit_page_deallocate_with_page_impl
that were not actually *deallocating*, but simply shrinking the object
or checking its size. This patch fixes that by guarding the retag
behind a switch as appropriate.

Canonical link: <a href="https://commits.webkit.org/303411@main">https://commits.webkit.org/303411@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d389008e3b2cf80118b1ec185fb9882f10882569

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132327 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43365 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139842 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/81410704-e027-4a3d-b5de-8263eedec5cb) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4581 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6b865824-19a2-4c47-8e00-022303e0be49) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135273 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118533 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/81951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/890653db-6d6f-4d4f-a810-503e445b0afd) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83067 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124395 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142493 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130835 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/4492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37236 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4574 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/109716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27794 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/3404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114805 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4546 "Failed to checkout and rebase branch from PR 54197") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163802 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/4378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/42559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/4503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_zero_memory.h:
```diff
@@ -35,9 +35,16 @@ PAS_BEGIN_EXTERN_C;
 static PAS_ALWAYS_INLINE void pas_zero_memory(void* memory, size_t size)
 {
     PAS_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    /*
+     * MTE systems perform poorly when zeroing large chunks
+     * of memory, so we set TCO when applicable to avoid that overhead.
+     */
+    if (PAS_USE_MTE)
+        PAS_MTE_CHECK_TAG_AND_SET_TCO(memory);
```
Doing it like this (rather than thru PAS_MTE_HANDLE) is necessary to break the dependency on pas_mte.h, as pas_zero_memory.h is very high up in the include graph and we want to use some of its descendants in pas_mte.h.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -926,6 +926,10 @@ void* pas_mte_system_heap_realloc_zero_tagged(malloc_zone_t* zone, void* ptr, si
         (void)ptr; \
         (void)size; \
+        if (PAS_USE_MTE) { \
+            if (PAS_MTE_SHOULD_TAG_PAGE(page_config)) \
```
Maybe combine these lines

Achierius replied:
Changed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_bitfit_page_inlines.h:
```diff
@@ -901,8 +899,15 @@ static PAS_ALWAYS_INLINE uintptr_t pas_bitfit_page_deallocate_with_page_impl(
-    PAS_PROFILE(BITFIT_PAGE_DEALLOCATION, page_config, begin, original_object_size);
-    PAS_MTE_HANDLE(BITFIT_PAGE_DEALLOCATION, page_config, begin, original_object_size);
+    switch (mode) {
+    case pas_bitfit_page_deallocate_with_page_impl_get_allocation_size_mode:
```
shouldn't do it for this mode

Achierius replied:
Fixed

Achierius committed:
[libpas] Extend Retag-on-Scavenge to cover bitfit objects as well
rdar://156723433
https://bugs.webkit.org/show_bug.cgi?id=300151

Reviewed by Dan Hecht.

This patch finishes the retag-on-scavenge work by extending the
implementation to cover bitfit objects as well.
The main issue with the prior implementation is that it would attempt
to retag even for calls to pas_bitfit_page_deallocate_with_page_impl
that were not actually *deallocating*, but simply shrinking the object
or checking its size. This patch fixes that by guarding the retag
behind a switch as appropriate.

Canonical link: https://commits.webkit.org/303411@main
